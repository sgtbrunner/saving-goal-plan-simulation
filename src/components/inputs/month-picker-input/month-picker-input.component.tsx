import { useState } from 'react';

import ArrowButton from '../../buttons/arrow-button/arrow-button.component';
import { MonthPickerInputProps } from './month-picker-input.types';
import {
  MonthPickerInputContainer,
  DateContainer,
  MonthParagraph,
  YearParagraph,
} from './month-picker-input.styles';
import { InputLabel } from '../../../styles/global.styles';

const MonthPickerInput = ({
  label,
  name = 'reachDate',
}: MonthPickerInputProps): JSX.Element => {
  const [reachDate, setReachDate] = useState({ month: 'November', year: 2021 });
  const { month, year } = reachDate;
  return (
    <div>
      <InputLabel htmlFor={name}>{label}</InputLabel>
      <MonthPickerInputContainer>
        <ArrowButton direction="left" aria-label="Previous month" />
        <DateContainer>
          <MonthParagraph>{month}</MonthParagraph>
          <YearParagraph>{year}</YearParagraph>
        </DateContainer>
        <ArrowButton direction="right" aria-label="Next month" />
      </MonthPickerInputContainer>
    </div>
  );
};

export default MonthPickerInput;
