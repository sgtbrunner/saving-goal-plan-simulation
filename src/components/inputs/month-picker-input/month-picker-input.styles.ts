import styled from 'styled-components';

import {
  SMALL_DEVICE_BREAKPOINT,
  InputContainer,
} from '../../../styles/global.styles';

export const MonthPickerInputContainer = styled(InputContainer)`
  justify-content: space-between;
  align-items: center;
  min-height: 66px;
  min-width: 188px;
`;

export const DateContainer = styled.div`
  font-size: var(--xxs-font-size);
  line-height: var(--sm-line-height);
  text-align: center;

  p {
    margin: 0;
  }

  @media (min-width: ${SMALL_DEVICE_BREAKPOINT}px) {
    font-size: var(--xs-font-size);
    line-height: var(--md-line-height);
  }
`;

export const MonthParagraph = styled.p`
  font-weight: var(--semibold-font-weight);
  color: var(--blue-gray-900-color);
`;

export const YearParagraph = styled.p`
  font-weight: var(--light-font-weight);
  color: var(--blue-gray-400-color);
`;
