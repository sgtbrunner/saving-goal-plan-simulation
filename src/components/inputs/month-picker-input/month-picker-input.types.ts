export type MonthPickerInputProps = {
  label: string;
  name?: string;
};

export interface MonthPickerInputState {
  step: number;
  year: string;
  month: string;
}
