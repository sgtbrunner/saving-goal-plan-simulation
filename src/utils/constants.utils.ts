// OPERATIONAL CONSTANTS
export const FIRST_STEP = 1;
export const LOCALE = 'en';
export const INITIAL_GOAL_AMOUNT = 25000;
export const PREVIOUS_BUTTON_ID = 'previous-month';
export const NEXT_BUTTON_ID = 'next-month';
export const ARROW_RIGHT = 'arrow-right';
export const ARROW_LEFT = 'arrow-left';

// GENERAL CONSTANTS
export const ID = 'id';
export const ORIGIN = 'Origin';
export const LEFT = 'left';
export const RIGHT = 'right';
export const WHITE = 'white';
export const DOLLAR_SIGN = '$';
