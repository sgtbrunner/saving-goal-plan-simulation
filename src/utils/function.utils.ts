export const clearInvalidCharacters = (value: string): string =>
  value.replace(/\D|^0+/g, '');

export const convertInputToNumber = (value: string): number => {
  if (value.includes('.')) {
    const [decimalPart, fractionalPart] = value.split('.');
    return Number(
      [
        clearInvalidCharacters(decimalPart),
        '.',
        clearInvalidCharacters(fractionalPart),
      ].join('')
    );
  }

  return Number(value.replace(/\D|^0+/g, ''));
};
