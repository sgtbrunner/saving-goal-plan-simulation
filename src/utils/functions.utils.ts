import { FIRST_STEP, LOCALE } from './constants.utils';
import { MonthPickerInputState } from './types.utils';

export const getDate = (customDate?: string | number): Date =>
  customDate ? new Date(customDate) : new Date();

export const monthFormatter = new Intl.DateTimeFormat(LOCALE, {
  month: 'long',
});

export const yearFormatter = new Intl.DateTimeFormat(LOCALE, {
  year: 'numeric',
});

export const getMonthPickerState = (
  step = FIRST_STEP,
  modifier = 0
): MonthPickerInputState => {
  const today = getDate();
  const targetMonthISO = today.setMonth(today.getMonth() + step + modifier, 1);

  return {
    step: step + modifier,
    year: yearFormatter.format(targetMonthISO),
    month: monthFormatter.format(targetMonthISO),
  };
};

export const getSignificantAndFractionalDigits = (
  value: string
): Array<string> => value.split('.');
