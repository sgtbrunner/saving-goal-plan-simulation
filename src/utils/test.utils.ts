import { INITIAL_GOAL_AMOUNT } from './constants.utils';
import { getDate, monthFormatter, yearFormatter } from './functions.utils';
import { SavingGoalState } from './types.utils';

// TAGS
export const ALT = 'alt';
export const BANNER = 'banner';
export const BUTTON = 'button';
export const IMG = 'img';
export const TEST = 'test';

// TEST IDs
export const CARD = 'card';
export const MAIN_COMPONENT = 'main-component';
export const MAIN_TITLE = 'main-title';
export const MONTHLY_AMOUNT = 'monthly-amount';
export const SUMMARY_TEXT = 'summary-text';

// FUNCTIONS
export const getDefaultSavingGoal = (): SavingGoalState => {
  const today = getDate();
  const nextMonth = today.setMonth(today.getMonth() + 1, 1);
  return {
    goalAmount: INITIAL_GOAL_AMOUNT,
    reachDateYear: yearFormatter.format(nextMonth),
    reachDateMonth: monthFormatter.format(nextMonth),
    monthlyAmount: INITIAL_GOAL_AMOUNT,
    monthlyDeposits: 1,
  };
};
