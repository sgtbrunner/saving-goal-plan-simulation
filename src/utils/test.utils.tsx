import React from 'react';
import { render, RenderResult } from '@testing-library/react';
import { Provider } from 'react-redux';
import { createStore } from 'redux';

import rootReducer from '../redux/root-reducer';
import { INITIAL_GOAL_AMOUNT } from './constants.utils';
import { getDate, monthFormatter, yearFormatter } from './functions.utils';
import { SavingGoalState } from './types.utils';

// TAG CONSTANTS
export const ALT = 'alt';
export const BANNER = 'banner';
export const BUTTON = 'button';
export const IMG = 'img';
export const TEST = 'test';
export const AMOUNT = 'amount';

// TEST ID CONSTANTS
export const CARD = 'card';
export const MAIN_COMPONENT = 'main-component';
export const MAIN_TITLE = 'main-title';
export const MONTHLY_AMOUNT = 'monthly-amount';
export const SUMMARY_TEXT = 'summary-text';
export const CURRENCY_INPUT = 'currency-input';

// FUNCTIONS
export const getDefaultSavingGoal = (): SavingGoalState => {
  const today = getDate();
  const nextMonth = today.setMonth(today.getMonth() + 1, 1);
  return {
    goalAmount: INITIAL_GOAL_AMOUNT,
    reachDateYear: yearFormatter.format(nextMonth),
    reachDateMonth: monthFormatter.format(nextMonth),
    monthlyAmount: INITIAL_GOAL_AMOUNT,
    monthlyDeposits: 1,
  };
};

// RENDERER
type WrapperPropsType = {
  children: React.ReactNode | React.ReactNode[];
};

export const renderWithRedux = (
  component: React.ReactElement,
  store = createStore(rootReducer)
): RenderResult => {
  const ComponentWrapper = (props: WrapperPropsType) => (
    <Provider store={store}>{props.children}</Provider>
  );

  return render(component, {
    wrapper: ComponentWrapper as React.ComponentType,
  });
};
