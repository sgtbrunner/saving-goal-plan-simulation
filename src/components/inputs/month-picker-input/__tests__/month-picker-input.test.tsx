import {
  fireEvent,
  screen,
  RenderResult,
  waitFor,
} from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import { BUTTON, renderWithRedux } from '../../../../utils/test.utils';
import { MonthPickerInputProps } from '../month-picker-input.types';
import MonthPickerInput from '../month-picker-input.component';
import {
  ID,
  NEXT_BUTTON_ID,
  PREVIOUS_BUTTON_ID,
  ARROW_LEFT,
  ARROW_RIGHT,
} from '../../../../utils/constants.utils';
import { getMonthPickerState } from '../../../../utils/functions.utils';

const defaultProps: MonthPickerInputProps = {
  label: 'Date',
  updateReachDate: jest.fn(),
};

const mountComponent = (newProps?: MonthPickerInputProps): RenderResult =>
  renderWithRedux(<MonthPickerInput {...defaultProps} {...newProps} />);

describe('MonthPickerInput component', () => {
  it('should match snapshot', () => {
    const { container } = mountComponent();

    expect(container).toMatchSnapshot();
  });

  it('should render the month picker input label', () => {
    mountComponent();

    const label = screen.getByText(defaultProps.label);
    expect(label).toBeInTheDocument();
  });

  it('should render 2 (two) arrow button components within the rendered component', () => {
    mountComponent();

    const buttons = screen.getAllByRole(BUTTON);
    expect(buttons).toHaveLength(2);
  });

  it('should render the default month picker input values', () => {
    mountComponent();
    const { month: monthValue, year: yearValue } = getMonthPickerState();

    const month = screen.getByText(monthValue);
    const year = screen.getByText(yearValue);

    expect(month).toBeInTheDocument();
    expect(year).toBeInTheDocument();
  });

  describe('user interaction events', () => {
    let previousButton: HTMLElement;
    let nextButton: HTMLElement;

    const ARROW_UP = 'ArrowUp';
    const ARROW_DOWN = 'ArrowDown';

    beforeEach(() => {
      mountComponent();

      previousButton = screen.getAllByRole(BUTTON)[0];
      expect(previousButton).toBeInTheDocument();
      expect(previousButton).toHaveProperty(ID, PREVIOUS_BUTTON_ID);

      nextButton = screen.getAllByRole(BUTTON)[1];
      expect(nextButton).toBeInTheDocument();
      expect(nextButton).toHaveProperty(ID, NEXT_BUTTON_ID);
    });

    afterEach(() => {
      jest.clearAllMocks();
    });

    it('previous month button should be disabled by default', () => {
      expect(previousButton).toBeDisabled();
    });

    it('date update function should not be called when clicking on disabled previous button', () => {
      userEvent.click(previousButton);
      expect(defaultProps.updateReachDate).not.toHaveBeenCalled();
    });

    it('should increment and decrement the date displayed when clicking on arrow buttons', () => {
      // CLICKS ON NEXT BUTTON
      userEvent.click(nextButton);
      waitFor(() => {
        expect(defaultProps.updateReachDate).toHaveBeenCalledTimes(1);
      });

      // ASSERTS FOR NEW DATE STATE
      const { month: monthValue, year: yearValue } = getMonthPickerState(1, 1);

      const month = screen.getByText(monthValue);
      const year = screen.getByText(yearValue);

      expect(month).toBeInTheDocument();
      expect(year).toBeInTheDocument();

      // CLICKS ON PREVIOUS BUTTON
      userEvent.click(previousButton);
      waitFor(() => {
        expect(defaultProps.updateReachDate).toHaveBeenCalledTimes(1);
      });

      // ASSERTS FOR DEFAULT STATE
      const { month: defaultMonthValue, year: defaultYearValue } =
        getMonthPickerState();

      const defaultMonth = screen.getByText(defaultMonthValue);
      const defaultYear = screen.getByText(defaultYearValue);

      expect(defaultMonth).toBeInTheDocument();
      expect(defaultYear).toBeInTheDocument();
    });

    it('should call for update date when pressing arrow buttons on keyboard', () => {
      nextButton.focus();
      // PRESS ARROW RIGHT KEYBOARD BUTTON
      fireEvent.keyDown(nextButton, { key: ARROW_RIGHT, code: ARROW_RIGHT });
      waitFor(() => {
        expect(defaultProps.updateReachDate).toHaveBeenCalledTimes(1);
      });

      // PRESS ARROW RIGHT KEYBOARD BUTTON
      fireEvent.keyDown(nextButton, { key: ARROW_LEFT, code: ARROW_LEFT });
      waitFor(() => {
        expect(defaultProps.updateReachDate).toHaveBeenCalledTimes(1);
      });
    });

    it('should NOT call for update date when any different key is pressed down', () => {
      nextButton.focus();
      fireEvent.keyDown(nextButton, { key: ARROW_UP, code: ARROW_UP });
      expect(defaultProps.updateReachDate).not.toHaveBeenCalled();

      fireEvent.keyDown(nextButton, { key: ARROW_DOWN, code: ARROW_DOWN });
      expect(defaultProps.updateReachDate).not.toHaveBeenCalled();
    });
  });
});
