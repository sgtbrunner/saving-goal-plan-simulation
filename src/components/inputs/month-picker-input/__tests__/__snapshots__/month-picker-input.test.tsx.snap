// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MonthPickerInput component should match snapshot 1`] = `
<div>
  <div>
    <label
      class="sc-gtsrHT dDJGpo"
      for="reachDate"
    >
      Date
    </label>
    <div
      class="sc-dlnjwi sc-hKFxyN ieTIfq kqBlkA"
    >
      <button
        class="sc-bdnxRM leqYLS"
        disabled=""
        id="previous-month"
        type="button"
      >
        <img
          alt="arrow-left"
          src="arrow-left.svg"
        />
      </button>
      <div
        class="sc-eCApnc gDrXuf"
      >
        <p
          class="sc-jSFjdj fCWTUk"
        >
          December
        </p>
        <p
          class="sc-gKAaRy bsozYs"
        >
          2021
        </p>
      </div>
      <button
        class="sc-bdnxRM leqYLS"
        id="next-month"
        type="button"
      >
        <img
          alt="arrow-right"
          src="arrow-right.svg"
        />
      </button>
    </div>
  </div>
</div>
`;
